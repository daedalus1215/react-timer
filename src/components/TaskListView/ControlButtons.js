// @TODO: Left off here moving the 3 buttons from TaskListView over into this component
// @TODO: Then we want to control when to display the other 2 buttons
// @TODO: Based on if there are any tasks. So we might be pushing Tasks into the context as well and
// @TODO: Creating a new context or adding to the existing one.
import React, { useEffect, useState } from "react";
import { NavLink } from "react-router-dom";
import PropTypes from "prop-types";
import "./TaskListView.css";
import JsonWriter from "./JsonWriter";
import useTaskEditContext from "../../Form/EditTask/useTaskEditContext";

import { getFormattedDate } from "../../utils/DateFormat";

const ControlButtons = ({ list, tasks, setTasks }) => {
  const { setMessage } = useTaskEditContext();

  const handleDownload = () => {
    fetch(`/api/tasks`)
      .then(response => response.json())
      .then(tasks => {
        const date = new Date();
        const dateFormatted = getFormattedDate(date);

        const timeTask = {
          date: dateFormatted
        };

        tasks.forEach(function(task) {
          task.time = (task.time / 1000 / 60 / 60).toFixed(2);

          list.forEach(function(existingTask) {
            if (task.contractId === existingTask.key) {
              task.contract = existingTask.contract;
              task.customer = existingTask.customer;
            }
          });
        });

        timeTask.WorkUnit = tasks;

        const writer = new JsonWriter();
        writer.write(timeTask);
      });
  };

  /**
   * Setting a timeout when reloading, because I think the reloading is happening too fast.
   */
  const handleDelete = e => {
    e.preventDefault();
    fetch(`/api/tasks`, {
      method: "DELETE"
    })
      .then(response => response.json())
      .then(
        setTimeout(() => {
          setTasks([]);
          setMessage("Successfully deleted all tasks");
        }, 500)
      );
  };

  const showDownloadButton = () => {
    if (tasks !== null) {
      return (
        <a href="#" className="button-download" onClick={handleDownload}>
          <span className="glyphicon glyphicon-download-alt mr-5px"></span>
          Download
        </a>
      );
    }
  };

  return (
    <div className="task-list__header">
      {tasks?.length && (
        <button type="a" className="button-delete" onClick={handleDelete}>
          <span className="glyphicon glyphicon-remove mr-5px"></span>
          Delete
        </button>
      )}
      {tasks?.length && (
        <button type="a" className="button-download" onClick={handleDownload}>
          <span className="glyphicon glyphicon-download-alt mr-5px"></span>
          Download
        </button>
      )}
      <NavLink to={"/task/-1"} className="button-add">
        <span className="glyphicon glyphicon-plus mr-5px" />
        New Task
      </NavLink>
    </div>
  );
};

export default ControlButtons;
