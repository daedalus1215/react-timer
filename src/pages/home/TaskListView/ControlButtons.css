.button-add,
.button-add:hover {
  background-color: #ff6358;
  color: white;
  border-radius: 4px;
  height: 1.5em;
  border: none;
  padding: 1em;
  text-decoration: none;
  text-align: center;
  width: 95px;
}

.button-add:hover {
  background-color: #a4413a;
}

.button-download,
.button-download:hover {
  display: flex;
  background-color: #3e80ed;
  color: white;
  border-radius: 4px;
  width: 90px;
  height: 1.5em;
  border: none;
  padding: 1em;
  text-align: center;
  text-decoration: none;
  margin-left: 0.5em;
}

.button-download:hover {
  background-color: #184693;
}

.button-delete,
.button-delete:hover {
  background-color: #d51923;
  color: white;
  border-radius: 4px;
  width: 75px;
  height: 1.5em;
  border: none;
  padding: 1em;
  text-align: center;
  text-decoration: none;
}

.button-delete:hover {
  background-color: #af111a;
}
