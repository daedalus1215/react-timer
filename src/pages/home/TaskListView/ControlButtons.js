// @TODO: Left off here moving the 3 buttons from TaskListView over into this component
// @TODO: Then we want to control when to display the other 2 buttons
// @TODO: Based on if there are any tasks. So we might be pushing Tasks into the context as well and
// @TODO: Creating a new context or adding to the existing one.
import React from 'react';
import { NavLink } from 'react-router-dom';
import { writeJsonFile } from './writeJsonFile';
import formatTimeContractAndCustomer from './formatTimeContractAndCustomer';
import useTaskEditContext from 'hooks/useTaskEditContext';
import { getFormattedDate } from 'utils';
import styles from './ControlButtons.module.css';

const ControlButtons = () => {
  const { setMessage, projects, tasks, updateTasks } = useTaskEditContext();

  const handleDownload = () => {
    const dateFormatted = getFormattedDate(new Date());

    const timeTask = {
      date: dateFormatted
    };

    timeTask.WorkUnit = tasks.map(task =>
      formatTimeContractAndCustomer(task, projects)
    );

    writeJsonFile(timeTask);
  };

  /**
   * Setting a timeout when reloading, because I think the reloading is happening too fast.
   */
  const handleDelete = async e => {
    e.preventDefault();
    await fetch(`/api/tasks`, {
      method: 'DELETE'
    })
      .then(response => response.json())
      .then(
        setTimeout(() => {
          updateTasks([]);
          setMessage('Successfully deleted all tasks');
        }, 500)
      );
  };

  return (
    <>
      {!tasks?.length || (
        <button
          type="a"
          className={styles.buttonDelete}
          onClick={handleDelete}
          data-test-id="btn-delete"
        >
          <span className="glyphicon glyphicon-remove mr-5px"></span>
          Delete
        </button>
      )}
      {!tasks?.length || (
        <button
          type="a"
          className={styles.buttonDownload}
          onClick={handleDownload}
          data-test-id="btn-download"
        >
          <span className="glyphicon glyphicon-download-alt mr-5px"></span>
          Download
        </button>
      )
      }
      <NavLink to={'/task/-1'} className={styles.buttonAdd} data-test-id="btn-new">
        <span className="glyphicon glyphicon- plus mr-5px" />
        New Task
    </NavLink >
    </>
  );
};

export default ControlButtons;
