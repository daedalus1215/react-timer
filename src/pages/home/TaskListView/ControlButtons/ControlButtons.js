import React from 'react';
import { useBrowserHistory } from 'hooks';
import DownloadButton from './DownloadButton/DownloadButton';
import NewButton from './NewButton/NewButton';
import styles from './ControlButtons.module.css';

const ControlButtons = tasks => {
  const { push } = useBrowserHistory();

  return (
    <div className={styles.controlButtons} data-test-id="control-buttons">
      <button className="glyphicon glyphicon-home" onClick={() => { push("/tag") }} />
      <DownloadButton tasks={tasks} />
      <NewButton />
    </div>
  )
};

export default ControlButtons;
