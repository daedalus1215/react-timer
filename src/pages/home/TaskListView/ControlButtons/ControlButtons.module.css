.buttonAdd,
.buttonAdd:hover {
  background-color: #ff6358;
  color: white;
  border-radius: 4px;
  height: 1.5em;
  border: none;
  padding: 1em;
  text-decoration: none;
  text-align: center;
  width: 95px;
}

.buttonAdd:hover {
  background-color: #a4413a;
}

