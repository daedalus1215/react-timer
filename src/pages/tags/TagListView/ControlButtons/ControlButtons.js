import React from 'react';
import { useBrowserHistory } from 'hooks';
import NewButton from './NewButton/NewButton';
import styles from './ControlButtons.module.css';

const ControlButtons = () => {
  const { push } = useBrowserHistory();
  return (
    <div className={styles.controlButtons} data-test-id="control-buttons">
      <button className="glyphicon glyphicon-home" onClick={() => { push("/") }} />
      <NewButton />
    </div>
  );
};

export default ControlButtons;
