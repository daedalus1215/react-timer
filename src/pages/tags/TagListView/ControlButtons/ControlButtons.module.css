.controlButtons {
  display: flex;
  height: 100%;
}
