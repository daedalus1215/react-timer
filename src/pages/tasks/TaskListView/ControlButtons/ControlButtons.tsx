import React from 'react';
import DownloadButton from './DownloadButton/DownloadButton';
import NewButton from './NewButton/NewButton';
import HomeButton from './HomeButton/HomeButton';
import UploadButton from './UploadButton/UploadButton';
import styles from './ControlButtons.module.css';
import { TaskInterface } from 'interfaces/pages/tasks/Task';

interface ControlButtonsProps {
  tasks: TaskInterface[];
}
const ControlButtons: React.FC<ControlButtonsProps> = ({tasks}) => {
  return (
  <div className={styles.controlButtons} data-testid="control-buttons">
    <HomeButton />
    <DownloadButton tasks={tasks} />
    <UploadButton />
    <NewButton />
  </div>
)};

export default ControlButtons;
