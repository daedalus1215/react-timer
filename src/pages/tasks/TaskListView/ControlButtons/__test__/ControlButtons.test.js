import { render } from '@testing-library/react'
import ControlButtons from '../ControlButtons';

jest.mock('../NewButton/NewButton', () => {
  return () => <div className="NewButton">NewButton</div>
});
jest.mock('../DownloadButton/DownloadButton', () => {
  return () => <div className="DownloadButton">DownloadButton</div>
});
jest.mock('../HomeButton/HomeButton', () => {
  return () => <div className="HomeButton">HomeButton</div>
});
jest.mock('../UploadButton/UploadButton', () => {
  return () => <div className="UploadButton">UploadButton</div>
});

describe('src/pages/tasks/TaskListView/ControlButtons/__test__/ControlButtons', () => {
  describe('ControlButtons', () => {
    it('should display ControlButtons', () => {
      // Arrange
      const tasks = [{}];

      // Act
      const target = render(<ControlButtons tasks={tasks} />);
      const actual = target.getByTestId('control-buttons');

      // Assert
      expect(actual).toBeInTheDocument();
      expect(actual).toMatchSnapshot();
    });
  });
});
