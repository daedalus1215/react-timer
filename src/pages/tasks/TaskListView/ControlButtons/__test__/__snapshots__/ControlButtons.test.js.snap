// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/pages/tasks/TaskListView/ControlButtons/__test__/ControlButtons ControlButtons should display ControlButtons 1`] = `
<div
  class="controlButtons"
  data-testid="control-buttons"
>
  <div
    class="HomeButton"
  >
    HomeButton
  </div>
  <div
    class="DownloadButton"
  >
    DownloadButton
  </div>
  <div
    class="UploadButton"
  >
    UploadButton
  </div>
  <div
    class="NewButton"
  >
    NewButton
  </div>
</div>
`;
